src/mlpack/core/optimizers/svrg/svrg_impl.hpp

```diff
@@ -0,0 +1,179 @@
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  size_t currentFunction = 0;
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Calculate the first objective function.
+  for (size_t i = 0; i < numFunctions; i += batchSize)
+  {
+    const size_t effectiveBatchSize = std::min(batchSize, numFunctions - i);
+    overallObjective += function.Evaluate(iterate, i, effectiveBatchSize);
+  }
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat fullGradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "SVRG: iteration " << i << ", objective " << overallObjective
+          << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "SVRG: converged to " << overallObjective
+            << "; terminating  with failure.  Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "SVRG: minimized within tolerance " << tolerance << "; "
+            << "terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        function.Shuffle();
+    }
+
+    // Compute the full gradient.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    function.Gradient(iterate, 0, fullGradient, effectiveBatchSize);
+    for (size_t f = effectiveBatchSize; f < numFunctions;
+        /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      fullGradient += gradient;
+
+      f += effectiveBatchSize;
+    }
+    fullGradient /= (double) numFunctions;
```

Comment:
It seems to me like the calculation of the full gradient here could be in the conditional `if ((currentFunction % numFunctions) == 0)`, since it seems with SVRG the full gradient is calculated at the start of an iteration.

Reply:
If we introduce the update frequency for the inner loop a user can control the full gradient calculation, what do you think? We can also remove the inner loop and put the full gradient inside the if condition.

Reply:
I see, I think that would be good to add.  For the default I guess the update frequency would be once per full pass over the data.